multicluster/controllers/multicluster/core/remote_cluster.go

```diff
@@ -0,0 +1,353 @@
+	if err := r.List(context.TODO(), memberAnnounceList, client.InNamespace(r.GetNamespace())); err != nil {
+		return err
+	}
+	var localClusterMemberAnnounce multiclusterv1alpha1.MemberClusterAnnounce
+	localClusterMemberAnnounceExist := false
+	if len(memberAnnounceList.Items) != 0 {
+		for _, memberAnnounce := range memberAnnounceList.Items {
+			if memberAnnounce.ClusterID == string(r.remoteClusterManager.GetLocalClusterID()) {
+				localClusterMemberAnnounceExist = true
+				localClusterMemberAnnounce = memberAnnounce
+				break
+			}
+		}
+	}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFunc()
+	if localClusterMemberAnnounceExist {
+		if r.remoteClusterManager.GetElectedLeaderClusterID() != common.INVALID_CLUSTER_ID {
+			localClusterMemberAnnounce.LeaderClusterID = string(r.remoteClusterManager.GetElectedLeaderClusterID())
+		}
+
+		if localClusterMemberAnnounce.Annotations == nil {
+			localClusterMemberAnnounce.Annotations = make(map[string]string)
+		}
+		localClusterMemberAnnounce.Annotations["touch-ts"] = time.Now().String()
+		// do an update
+		if err := r.Update(ctx, &localClusterMemberAnnounce, &client.UpdateOptions{}); err != nil {
+			klog.Error(err, "Failed to update member announce", "cluster", r.GetClusterID())
+			return err
+		}
+	} else {
+		// create happens first before a leader can be elected, when the create is successful
+		// it marks the connectivity status to leader election can occur
+		// Therefore the first create will not populate the leader cluster id
+		localClusterMemberAnnounce.ClusterID = string(r.remoteClusterManager.GetLocalClusterID())
+		localClusterMemberAnnounce.Name = "member-announce-from-" + string(r.remoteClusterManager.GetLocalClusterID())
+		localClusterMemberAnnounce.Namespace = r.Namespace
+		localClusterMemberAnnounce.ClusterSetID = string(r.ClusterSetID)
+		if err := r.Create(ctx, &localClusterMemberAnnounce, &client.CreateOptions{}); err != nil {
+			klog.Error(err, "Failed to create member announce", "cluster", r.GetClusterID())
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *remoteCluster) updateRemoteClusterStatus(connected bool, err error) {
+	if r.connected == connected {
+		return
+	}
+
+	klog.Infof("Updating remote cluster status", "cluster", r.GetClusterID(), "connected", connected)
+
+	// TODO: Tolerate transient failures so we dont oscillate between connected and disconnected.
+	r.connected = connected
+}
+
+/**
+ * ---------------------------
+ * commonArea Implementation
+ * ---------------------------
+ */
+
+func (r *remoteCluster) GetClusterID() common.ClusterID {
+	return r.ClusterID
+}
+
+func (r *remoteCluster) GetNamespace() string {
+	return r.Namespace
+}
+
+/**
+ * ---------------------------
+ * RemoteCluster Implementation
+ * ---------------------------
+ */
+
+/**
+ * Once connected to the remote cluster, the start method run a timer
+ * on a go routine to periodically write MemberAnnounce CRD into the remote
```

Comment:
Why we need to update MemberAnnounce periodically? In the latest design, does member cluster still exports to all leader clusters? This seems introduce unnecessary overhead. For simplicity, I still feel we can just have a single active leader for the whole ClusterSet. But let us revisit this later.

Reply:
So, if member writes MemberClusterAnnounce only once, it is possible for the member to know when it is connected or not connected to the leader. But the leader has no way to know that, e.g. if it needs to cleanup some resources from the ImportResource computation, it cannot know that. Thats why I implemented member cluster to write periodically, so leader can see if a member stops writing it means its no longer connected and leader can do some cleanup like remove all its ExportResources from the ImportResource computation etc.

Reply:
How long after disconnection we clean up ResourceExport? Should we clean up only when ClusterClaim is removed?

Reply:
I dont think that part is implemented yet. But the idea is not to clean up, but mark the status for the Member as Ready=False which will trigger the ResourceExport controller in leader cluster to remove those ResourceExports from the computation. @luolanzone may know more details.
But yes we will cleanup resources from member only when the member is deboarded from the ClusterSet

Reply:
we didn't handle the case in resource exchange pipeline to delete resource when member is disconnected. in common area implementation when member is disconnected, it only means member is disconnected with leader, so the connection between members might be still healthy. I am not sure if there is a good way to handle such case @jianjuns  any suggestion? I think we may handle deletion more carefully.
for the off-boarding case, we definitely need to clean up the resource, this part is missing, I will check it.

Reply:
> but mark the status for the Member as Ready=False which will trigger the ResourceExport controller in leader cluster to remove those ResourceExports from the computation.
What computation you mean here? I feel we should not remove Endpoints from disconnected clusters. Probably @luolanzone means the same.